pkg/fx/retrieval/ucan/handlers/provider.go

```diff
@@ -1,16 +1,82 @@
 package handlers
 
 import (
+	"context"
+
 	"go.uber.org/fx"
 
+	logging "github.com/ipfs/go-log/v2"
+	"github.com/storacha/go-libstoracha/capabilities/space/content"
+	"github.com/storacha/go-ucanto/core/invocation"
+	"github.com/storacha/go-ucanto/core/receipt"
+	fdm "github.com/storacha/go-ucanto/core/result/failure/datamodel"
+	ucanserver "github.com/storacha/go-ucanto/server"
+	ucanretrieval "github.com/storacha/go-ucanto/server/retrieval"
+
+	"github.com/storacha/piri/pkg/service/egresstracking"
 	"github.com/storacha/piri/pkg/service/retrieval/ucan"
 )
 
+var log = logging.Logger("retrieval/ucan")
+
 var Module = fx.Module("retrieval/ucan/handlers",
 	fx.Provide(
 		fx.Annotate(
 			ucan.SpaceContentRetrieve,
 			fx.ResultTags(`group:"ucan_retrieval_options"`),
 		),
+		fx.Annotate(
+			withErrorHandler,
+			fx.ResultTags(`group:"ucan_retrieval_options"`),
+		),
+		fx.Annotate(
+			withReceiptLogger,
+			fx.ResultTags(`group:"ucan_retrieval_options"`),
+		),
 	),
 )
+
+func withErrorHandler() ucanretrieval.Option {
+	return ucanretrieval.WithErrorHandler(func(err ucanserver.HandlerExecutionError[any]) {
+		l := log.With("error", err.Error())
+		if s := err.Stack(); s != "" {
+			l = l.With("stack", s)
+		}
+		l.Error("ucan retrieval handler execution error")
+	})
+}
+
+func withReceiptLogger(ets *egresstracking.EgressTrackingService) ucanretrieval.Option {
+	return ucanretrieval.WithReceiptLogger(func(_ context.Context, rcpt receipt.AnyReceipt, inv invocation.Invocation) {
+		// Egress Tracking is optional, the service will be nil if it is disabled
+		if ets == nil {
+			return
+		}
+
+		// Collect the receipt in a goroutine to avoid blocking the handler
+		go func() {
+			// Make sure the receipt is self-contained, i.e. it also has invocation blocks
+			fullRcpt, err := rcpt.Clone()
+			if err != nil {
+				log.Errorw("cloning receipt", "error", err)
+				return
+			}
+
+			if err := fullRcpt.AttachInvocation(inv); err != nil {
+				log.Errorw("attaching invocation to receipt", "error", err)
+				return
+			}
+
+			retrievalRcpt, err := receipt.Rebind[content.RetrieveOk, fdm.FailureModel](fullRcpt, content.RetrieveOkType(), fdm.FailureType())
+			if err != nil {
+				log.Errorw("rebinding receipt", "error", err)
+				return
+			}
+
+			if err := ets.AddReceipt(context.Background(), retrievalRcpt); err != nil {
+				log.Errorw("adding receipt to egress tracking service", "error", err)
+				return
+			}
+		}()
```

Comment:
I liked this better when we were passing the context from the caller here 😄 can we add it back?

Reply:
I agree with your concerns. A bounded pool or any other rate limiting mechanism would be the way to ensure there is some back-pressure on clients, but happy to remove the goroutine in the meantime.

About making the request fail if adding the receipt fails, there is a problem with the current implementation. `AddReceipt` enqueues a task in a jobqueue if there is a new batch to send, so this is happening asynchronously. If we think the client should not get its content delivered if recording the receipt fails (which is a fair point), then we shouldn't use the jobqueue and just implement a synchronous version of the task.

I created https://github.com/storacha/piri/issues/263 to help keeping this in mind.

Reply:
For now lets remove the go routine.

As far as receipt recording failures:
- If the node is unable to `Append` a receipt to the store, due to lack of space, failure to rotate, etc. that's when I'd want to abort serving the data of the request. Since the node is unable to record a record of the content it served for payment. 
- Failing to report a batch to the egress service is also not ideal, but I view as a less critical error since a record of the content served will still exist on the filesystem, and manual intervention could be taken to "re-report" it for payment.